python/lsst/daf/butler/datastores/chainedDatastore.py

```diff
@@ -0,0 +1,342 @@
+        ------
+        NotImplementedError
+            If all chained datastores have no ingest implemented or if
+            a transfer mode of `None` is specified.
+        """
+        log.debug("Ingesting %s (transfer=%s)", args[1], kwargs["transfer"])
+
+        if kwargs["transfer"] is None:
+            raise NotImplementedError("ChainedDatastore does not support transfer=None")
+
+        counter = 0
+        for datastore in self.datastores:
+            try:
+                datastore.ingest(*args, **kwargs)
+            except NotImplementedError:
+                counter += 1
+                pass
+        if counter == len(self.datastores):
+            raise NotImplementedError("Ingest not implemented by any of the chained datastores")
+
+    def getUri(self, ref, predict=False):
+        """URI to the Dataset.
+
+        The returned URI is from the first datastore in the list that has
+        the dataset. If no datastores have the dataset and prediction is
+        allowed, the predicted URI for the first datastore in the list will
+        be returned.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+        predict : `bool`
+            If `True`, allow URIs to be returned of datasets that have not
+            been written.
+
+        Returns
+        -------
+        uri : `str`
+            URI string pointing to the Dataset within the datastore. If the
+            Dataset does not exist in the datastore, and if ``predict`` is
+            `True`, the URI will be a prediction and will include a URI
+            fragment "#predicted".
+
+        Notes
+        -----
+        If the datastore does not have entities that relate well
+        to the concept of a URI the returned URI string will be
+        descriptive. The returned URI is not guaranteed to be obtainable.
+
+        Raises
+        ------
+        FileNotFoundError
+            A URI has been requested for a dataset that does not exist and
+            guessing is not allowed.
+        """
+        predicted = None
+        for datastore in self.datastores:
+            if datastore.exists(ref):
+                return datastore.getUri(ref)
+            elif predicted is None and predict:
+                predicted = datastore.getUri(ref, predict)
+
+        if predicted is not None:
+            return predicted
+
+        raise FileNotFoundError("Dataset {} not in any datastore".format(ref))
+
+    def remove(self, ref):
+        """Indicate to the Datastore that a Dataset can be removed.
+
+        The dataset will be removed from each datastore.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+
+        Raises
+        ------
+        FileNotFoundError
+            Attempt to remove a dataset that does not exist.
+        """
+        for datastore in self.datastores:
+            datastore.remove(ref)
```

Comment:
Given that `get` works if only one `Datastore` has the dataset, this should probably behave like `ingest` and raise only if it wasn't removed anywhere.

Reply:
Oh yes. That shows that my tests need to be extended.